src/cc/libbpf.c

```diff
@@ -400,43 +426,78 @@ int bpf_prog_load(enum bpf_prog_type prog_type, const char *name,
     }
   }
 
-  if (ret < 0 && !log_buf) {
+  if (ret < 0) {
+    if (log_buf) {
+      // If user alraedy provided log buffer, just print message and return.
+      bpf_print_hints(ret, log_buf);
+      if (errno == ENOSPC)
+        fprintf(stderr, "bpf: log_buf size may be insufficient\n");
+      goto return_result;
+    }
 
-    buffer_size = LOG_BUF_SIZE;
-    // caller did not specify log_buf but failure should be printed,
-    // so repeat the syscall and print the result to stderr
+    // User did not provide log buffer. We will try to increase size of
+    // our temporary log buffer to get full error message.
+    if (tmp_log_buf)
+      free(tmp_log_buf);
+    tmp_log_buf_size = LOG_BUF_SIZE;
+    attr.log_level = 1;
```

Comment:
One more thing. Do we want integer log_level or boolean do_logging? Providing log_level is more flexible for users as log_level=2 can print out more info than log_level=1 in kernel. Maybe the above line needs to inherit user log_level?

Reply:
Good point. `log_level` is already integer, I will do another patch to add a new flag to do `log_level=2` for more detailed register state inspection.

Reply:
If user_buf != NULL, we will have 
   if (input_log_level == 0)
       log_level = 1
  else
      log_level = input_log_level

Can we do the same thing here?
attr.log_level = (log_level == 0) ? 1 : log_level.

This will cover log_level=2 case for C++ API, where people are more expert...
For Python side, we have log_level=1 based on debug flag settting, I feel this is good enough.
log_level=2 prints reg state for every insn, this is mostly useful to track verifier bug and often happens with bigger programs (C++ side) than python side.

Reply:
I thing enabling `log_level` based on that user provided `log_buf` makes the logic is a bit messy. In brief words, the implemented logic in this patch is:
- If user specified positive `log_level`, use the value for `log_level`.
- Otherwise if load failed, set `log_level` to be 1.
- When `log_level` enabled due to any above case:
  - If user provided `log_buf`, use user's `log_buf`.
  - Otherwise allocate use our temporary log buffer.

Do you feel that's a reasonable logic?

Reply:
This sounds okay, even with "Otherwise if load failed, set log_level to be 1". This is probably what we did earlier. Could you add this brief algorithm as the comments?

Reply:
@yonghong-song Yeah I was trying to keep current behavior mostly unchanged. Documentation is added on `libbpf.h`, do you have any suggestion on how to make it better?